detectors/azure/azurevm/vm.go

```diff
@@ -0,0 +1,142 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package azurevm // import "go.opentelemetry.io/contrib/detectors/azure/azurevm"
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
+)
+
+const (
+	defaultAzureVMMetadataEndpoint = "http://169.254.169.254/metadata/instance/compute?api-version=2021-12-13&format=json"
+)
+
+type config struct {
+	endpoint string
+}
+
+func newConfig(options ...Option) config {
+	c := config{defaultAzureVMMetadataEndpoint}
+	for _, option := range options {
+		c = option.apply(c)
+	}
+
+	return c
+}
+
+// Option applies an Azure VM detector configuration option.
+type Option interface {
+	apply(config) config
+}
+
+type optionFunc func(config) config
+
+func (fn optionFunc) apply(c config) config {
+	return fn(c)
+}
+
+// WithEndpoint sets the endpoint for obtaining a Azure instance metadata JSON.
+func WithEndpoint(e string) Option {
+	return optionFunc(func(c config) config {
+		c.endpoint = e
+
+		return c
+	})
+}
+
+type resourceDetector struct {
+	endpoint string
+}
+
+type vmMetadata struct {
+	VMId       *string `json:"vmId"`
+	Location   *string `json:"location"`
+	ResourceId *string `json:"resourceId"`
+	Name       *string `json:"name"`
+	VMSize     *string `json:"vmSize"`
+	OsType     *string `json:"osType"`
+	Version    *string `json:"version"`
+}
+
+// New returns a [resource.Detector] that will detect Azure VM resources.
+func New(opts ...Option) resource.Detector {
+	c := newConfig(opts...)
+	return &resourceDetector{c.endpoint}
+}
+
+// Detect detects associated resources when running on an Azure VM.
+func (detector *resourceDetector) Detect(ctx context.Context) (*resource.Resource, error) {
+	jsonMetadata, err := detector.getJSONMetadata()
+	if err != nil {
```

Comment:
This is compliant with the spec (thanks!), but might hide real errors when someone is running on an azure vm.  Should we look for a more specific error before returning the empty resource?  E.g. if the metadata endpoint returns a 500, that probably means we are running on an azure VM, and we should return that error.

Reply:
I restructured the code a bit so that an empty resource is returned for HTTP client errors (4* status codes, which means we couldn't reach the server and can't be sure we're running in Azure). For all other cases, an error is returned (we can be quite confident we're running in Azure, and something went wrong).

Reply:
lgtm